### ISSUE TITLE
"x.py check" shows error twice

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#76822]
I just got this output from `./x.py check`:
```
error[E0277]: the trait bound `&rustc_hir::def_id::DefId: rustc_middle::ty::query::sealed::IntoQueryParam<rustc_hir::def_id::DefId>` is not satisfied
   --> compiler/rustc_mir/src/transform/promote_consts.rs:748:53
    |
748 |                 if !self.tcx.is_promotable_const_fn(def_id) {
    |                                                     ^^^^^^ the trait `rustc_middle::ty::query::sealed::IntoQueryParam<rustc_hir::def_id::DefId>` is not implemented for `&rustc_hir::def_id::DefId`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: could not compile `rustc_mir`.

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error[E0277]: the trait bound `&rustc_hir::def_id::DefId: rustc_middle::ty::query::sealed::IntoQueryParam<rustc_hir::def_id::DefId>` is not satisfied
   --> compiler/rustc_mir/src/transform/promote_consts.rs:748:53
    |
748 |                 if !self.tcx.is_promotable_const_fn(def_id) {
    |                                                     ^^^^^^ the trait `rustc_middle::ty::query::sealed::IntoQueryParam<rustc_hir::def_id::DefId>` is not implemented for `&rustc_hir::def_id::DefId`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: build failed
```
I also noticed that the cargo "status bar" will show things like `rustc_mir, rustc_mir` or `rustc_driver, rustc_driver`. Is it building everything twice?

Cc @Mark-Simulacrum

**Comment by ehuss** [comment]
This is likely a consequence of #76258. Cargo checks the library and unit test in parallel, which can result in duplicate messages.  This is a cargo issue (tracked in https://github.com/rust-lang/cargo/issues/5128).  I'm thinking about adding a hash to track duplicates, but it will likely be a little tricky.

**Comment by RalfJung** [comment]
Ah that makes sense... closing in favor of https://github.com/rust-lang/cargo/issues/5128 then.

**Comment by Mark-Simulacrum** [comment]
FWIW we could "fix" this by running *first* the regular check and then the all-target check, but that would likely be a slowdown on most systems as you lose a bunch of potential parallelism.
